src/sentry/debug_files/artifact_bundles.py

```diff
@@ -34,14 +50,374 @@
+            return False
+
+    def _merge(
+        self, artifact_bundle: ArtifactBundle, archive: ArtifactBundleArchive, index_urls: bool
+    ):
+        bundle_index = self._update_bundles(artifact_bundle)
+
+        if index_urls:
+            self._iterate_over_urls(bundle_index, archive)
+        else:
+            self._iterate_over_debug_ids(bundle_index, archive)
+
+    def _update_bundles(self, artifact_bundle: ArtifactBundle) -> int:
+        if artifact_bundle.date_last_modified is None:
+            raise Exception("Can't index a bundle with no date last modified")
+
+        bundle_meta = BundleMeta(
+            id=artifact_bundle.id, timestamp=artifact_bundle.date_last_modified
+        )
+
+        bundle_index = self._index_of_bundle(artifact_bundle.id)
+        if bundle_index is None:
+            self._bundles.append(bundle_meta)
+            return len(self._bundles) - 1
+        else:
+            self._bundles[bundle_index] = bundle_meta
+            return bundle_index
+
+    def _iterate_over_debug_ids(self, bundle_index: int, archive: ArtifactBundleArchive):
+        for debug_id, _ in archive.get_all_debug_ids():
+            self._add_sorted_entry(self._files_by_debug_id, debug_id, bundle_index)
+
+    def _iterate_over_urls(self, bundle_index: int, archive: ArtifactBundleArchive):
+        for url in archive.get_all_urls():
+            self._add_sorted_entry(self._files_by_url, url, bundle_index)
+
+    def _add_sorted_entry(self, collection: Dict[T, List[int]], key: T, bundle_index: int):
+        entries = collection.get(key, [])
+        entries.append(bundle_index)
+        entries.sort(key=lambda index: self._bundles[index]["timestamp"], reverse=True)
+        collection[key] = entries
+
+    def remove(self, artifact_bundle_id: int) -> bool:
+        try:
+            self._remove(artifact_bundle_id)
+            return True
+        except Exception as e:
+            sentry_sdk.capture_exception(e)
+            metrics.incr("artifact_bundle_flat_file_indexing.error_when_removing_bundle")
+            return False
+
+    def _remove(self, artifact_bundle_id: int):
+        bundle_index = self._index_of_bundle(artifact_bundle_id)
+        if bundle_index is not None:
+            self._files_by_url = self._update_bundle_references(self._files_by_url, bundle_index)
+            self._files_by_debug_id = self._update_bundle_references(
+                self._files_by_debug_id, bundle_index
+            )
+            self._bundles.pop(bundle_index)
+        else:
+            raise Exception("The bundle you want do delete doesn't exist in the index.")
+
+    def _update_bundle_references(self, collection: Dict[T, List[int]], removed_bundle_index: int):
+        updated_collection: Dict[T, List[int]] = {}
+
+        for key, indexes in collection.items():
+            updated_indexes = [
+                self._compute_updated_index(index, removed_bundle_index)
```

Comment:
```suggestion
                index if index < removed_bundle_index else index - 1
```

IMO its easier to read with the code inlined, otherwise a comment will make this clearer.